File: String/Basics/Strings.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

func main() {
	// 1. Basic String Declaration and Printing
	myStr := "Hello World"
	fmt.Println(myStr) // Output: Hello World

	// 2. Accessing Characters in a String
	fmt.Println(string(myStr[0])) // H
	fmt.Println(string(myStr[1])) // e
	fmt.Println(string(myStr[len(myStr)-1])) // d

	// 3. String Concatenation
	str1 := "Hello"
	str2 := "World"
	fmt.Println(str1 + " " + str2) // Hello World

	// 4. String Repetition (Using a loop since Go doesn't have a direct string repetition operator)
	fmt.Println(strings.Repeat(str1, 3)) // HelloHelloHello
	fmt.Println(strings.Repeat(str2, 2)) // WorldWorld

	// 5. Slice a String
	fmt.Println(myStr[0:5]) // Hello
	fmt.Println(myStr[6:])  // World
	fmt.Printf("%T\n", myStr[0:3]) // string

	// 6. String Length
	fmt.Println(len(myStr)) // 11

	// 7. String Membership (Using strings.Contains)
	fmt.Println(strings.Contains(myStr, "H")) // true
	fmt.Println(strings.Contains(myStr, "z")) // false

	// 8. String Iteration
	for _, s := range myStr {
		fmt.Printf("%c ", s) // H e l l o   W o r l d
	}
	fmt.Println()

	// 9. String Comparison
	str3 := "World"
	fmt.Println(str1 == str2) // true
	fmt.Println(str1 == str3) // false

	// 10. String Formatting
	name := "Alice"
	age := 25
	fmt.Printf("Hello, my name is %s and I am %d years old.\n", name, age) // Hello, my name is Alice and I am 25 years old.
	
	// 11. String Formatting using fmt.Sprintf
	s := fmt.Sprintf("Hello, my name is %s and I am %d years old.", name, age)
	fmt.Println(s) // Hello, my name is Alice and I am 25 years old.

	// 12. String Formatting using fmt.Printf with f-string-like behavior (Go 1.13+)
	fmt.Printf("Hello, my name is %s and I am %d years old.\n", name, age) // Hello, my name is Alice and I am 25 years old.

	// 13. String Methods (Using strings package)
	myStrLower := "hello world"
	fmt.Println(strings.Title(myStrLower)) // Hello World
	fmt.Println(strings.ToUpper(myStrLower)) // HELLO WORLD
	fmt.Println(strings.ToLower(myStrLower)) // hello world
	fmt.Println(strings.Title(myStrLower)) // Hello World
	fmt.Println(strings.ToTitle(strings.ToLower(myStrLower))) // Hello World
	fmt.Println(strings.Count(myStrLower, "l")) // 3
	fmt.Println(strings.Index(myStrLower, "l")) // 2
	fmt.Println(strings.Index(myStrLower, "z")) // -1
	fmt.Println(strings.LastIndex(myStrLower, "l")) // 9
	// fmt.Println(strings.Index(myStrLower, "z")) // -1 (Would cause runtime error if used with Index)
	fmt.Println(strings.ReplaceAll(myStrLower, "world", "Alice")) // hello Alice
	fmt.Println(strings.HasPrefix(myStrLower, "hello")) // true
	fmt.Println(strings.HasSuffix(myStrLower, "world")) // true
	fmt.Println(strings.Split(myStrLower, " ")) // [hello world]
	fmt.Println(strings.Split(myStrLower, "l")) // [he "" o wor d]
	fmt.Println(strings.TrimSpace("   hello world   ")) // hello world
	fmt.Println(strings.TrimLeft("   hello world   ", " ")) // "hello world   "
	fmt.Println(strings.TrimRight("   hello world   ", " ")) // "   hello world"
	fmt.Println(unicode.IsLetter(rune(myStrLower[0]))) // true
	fmt.Println(unicode.IsDigit(rune(myStrLower[0]))) // false
	fmt.Println(unicode.IsLower(rune(myStrLower[0]))) // true
	fmt.Println(unicode.IsUpper(rune(myStrLower[0]))) // false
	fmt.Println(unicode.IsSpace(rune(' '))) // true

	// 14. String Join
	fmt.Println(strings.Join(strings.Split(myStr, ""), " ")) // H e l l o   W o r l d
	reversed := func(s string) string {
		runes := []rune(s)
		for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
			runes[i], runes[j] = runes[j], runes[i]
		}
		return string(runes)
	}
	fmt.Println(reversed(myStr)) // dlroW olleH

	// 15. String Reverse
	fmt.Println(reverseString(myStr)) // dlroW olleH

	// 16. String Palindrome
	str1Palindrome := "madam"
	fmt.Println(str1Palindrome == reverseString(str1Palindrome)) // true

	// 17. String to Integer
	numStr := "123"
	num, err := strconv.Atoi(numStr)
	if err != nil {
		fmt.Println("Error converting string to int:", err)
	} else {
		fmt.Println(num) // 123
		fmt.Printf("%T\n", num) // int
	}

	// 18. Integer to String
	num = 123
	numStr = strconv.Itoa(num)
	fmt.Println(numStr) // 123
	fmt.Printf("%T\n", numStr) // string

	// 19. String to List (Slice of runes)
	myStrList := []rune(myStr)
	fmt.Println(myStrList) // [72 101 108 108 111 32 87 111 114 108 100]

	// 20. List to String
	l := []rune{'h', 'e', 'l', 'l', 'o'}
	fmt.Println(string(l)) // hello

	// 21. String to Set (Using map for set-like behavior)
	myStrSet := make(map[rune]bool)
	for _, char := range myStr {
		myStrSet[char] = true
	}
	fmt.Println(myStrSet) // map[32:true 68:true 72:true 87:true 101:true 108:true 111:true 114:true 100:true]

	// 22. Set to String
	runes := []rune{'o', 'e', 'h', 'l'}
	fmt.Println(string(runes)) // oehl

	// 23. String to Dictionary (Using map)
	myStrDict := make(map[int]rune)
	for i, char := range myStr {
		myStrDict[i] = char
	}
	fmt.Println(myStrDict) // map[0:72 1:101 2:108 3:108 4:111 5:32 6:87 7:111 8:114 9:108 10:100]

	// 24. Dictionary to String
	d := map[int]string{
		0:  "h",
		1:  "e",
		2:  "l",
		3:  "l",
		4:  "o",
	}
	var dictStr []string
	for i := 0; i < len(d); i++ {
		dictStr = append(dictStr, d[i])
	}
	fmt.Println(strings.Join(dictStr, "")) // hello

	// 25. String to Tuple (Using slice of runes)
	myStrTuple := []rune(myStr)
	fmt.Println(myStrTuple) // [72 101 108 108 111 32 87 111 114 108 100]

	// 26. Tuple to String
	t := []rune{'h', 'e', 'l', 'l', 'o'}
	fmt.Println(string(t)) // hello
}

// Helper function to reverse a string
func reverseString(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

File: String/Basics/Strings.py
my_str = "Hello World"
print(my_str) # Hello World

# Accessing characters in a string
print(my_str[0]) # H
print(my_str[1]) # e
print(my_str[-1]) # d

# String Concatenation
str1 = "Hello"
str2 = "World"
print(str1 + " " + str2) # Hello World

# String Repetition
print(str1 * 3) # HelloHelloHello
print(str2 * 2) # WorldWorld

# Slice a string
print(my_str[0:5]) # Hello
print(my_str[6:]) # World
print(type(my_str[0:3])) # <class 'str'>

# String Length
print(len(my_str)) # 11

# String Membership
print('H' in my_str) # True
print('z' in my_str) # False

# String Iteration 
for s in my_str:
    print(s, end=" ") # H e l l o   W o r l d
print()

# String Comparison
str1 = "Hello"
str2 = "Hello"
str3 = "World"
print(str1 == str2) # True
print(str1 == str3) # False

# String Formatting
name = "Alice"
age = 25
print("Hello, my name is %s and I am %d years old." % (name, age)) # Hello, my name is Alice and I am 25 years old. 

# String Formatting using format method
print("Hello, my name is {} and I am {} years old.".format(name, age)) # Hello, my name is Alice and I am 25 years old.

# String Formatting using f-string
print(f"Hello, my name is {name} and I am {age} years old.") # Hello, my name is Alice and I am 25 years old.

# String Methods
my_str = "hello world"
print(my_str.capitalize()) # Hello world
print(my_str.upper()) # HELLO WORLD
print(my_str.lower()) # hello world
print(my_str.title()) # Hello World
print(my_str.swapcase()) # HELLO WORLD
print(my_str.count('l')) # 3
print(my_str.find('l')) # 2
print(my_str.find('z')) # -1
print(my_str.index('l')) # 2
# print(my_str.index('z')) # ValueError: substring not found
print(my_str.replace('world', 'Alice')) # hello Alice  
print(my_str.startswith('hello')) # True
print(my_str.endswith('world')) # True
print(my_str.split()) # ['hello', 'world']
print(my_str.split('l')) # ['he', '', 'o wor', 'd']
print(my_str.strip()) # hello world
print(my_str.lstrip()) # hello world
print(my_str.rstrip()) # hello world
print(my_str.isalnum()) # False
print(my_str.isalpha()) # False
print(my_str.isdigit()) # False
print(my_str.islower()) # True
print(my_str.isupper()) # False
print(my_str.isspace()) # False

# String Join
my_str = "Hello World"
print(" ".join(my_str)) # H e l l o   W o r l d
print("".join(reversed(my_str))) # dlroW olleH

# String Reverse
print(my_str[::-1]) # dlroW olleH
print("".join(reversed(my_str))) # dlroW olleH

# String Palindrome
str_1 = "madam"
print(str_1 == str_1[::-1]) # True

# String to Integer
num_str = "123"
print(int(num_str)) # 123
print(type(int(num_str))) # <class 'int'>

# Integer to String
num = 123
print(str(num)) # 123
print(type(str(num))) # <class 'str'>


# String to List
my_str = "hello"
print(list(my_str)) # ['h', 'e', 'l', 'l', 'o']

# List to String
l = ['h', 'e', 'l', 'l', 'o']
print("".join(l)) # hello

# String to Set
my_str = "hello"
print(set(my_str)) # {'o', 'e', 'h', 'l'}

# Set to String
s = {'o', 'e', 'h', 'l'}
print("".join(s)) # oehl

# String to Dictionary
my_str = "hello"
print(dict(enumerate(my_str))) # {0: 'h', 1: 'e', 2: 'l', 3: 'l', 4: 'o'}

# Dictionary to String
d = {0: 'h', 1: 'e', 2: 'l', 3: 'l', 4: 'o'}
print("".join(d.values())) # hello

# String to Tuple
my_str = "hello"
print(tuple(my_str)) # ('h', 'e', 'l', 'l', 'o')

# Tuple to String
t = ('h', 'e', 'l', 'l', 'o')
print("".join(t)) # hello